### PR TITLE
Added onion version runtime checks.

### PR DESCRIPTION
**Pull request by davidmoreno**
* Adds cmake setting the current version at version.h
* Functions to get each component and the full version
* Functions to test current version is ok to use
* Tests

### CODE REVIEW

File: git-version-gen [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 # usage: git-version-gen [srcdir]
 
-DEF_VER="v0.7.GIT"
+DEF_VER="v0.8.GIT"
```
Comment (zacharygrafton) [inline git-version-gen]: This is causing the CI build. Probably because .GIT won't match the REGEX in the cmake script.

**Comment by davidmoreno** [comment]
Thanks Zach, it was spot on. :)

I thinking of adding a onion_version_is_compatible_or_abort that does the abort if its not the proper version, as a easy bail out of incompatible versions, what do you think?

**Comment by zacharygrafton** [comment]
I wouldn't be opposed to the onion_version_is_compatible_or_abort, but I think, depending on how we choose to abort, that we could run into issues in C++ land, but that is easily remedied by adding some documentation.

**Comment by davidmoreno** [comment]
In the C++ bindings we could just use the onion_version_is_compatible and then throw an exception. No need to use the abort version.

**Comment by davidmoreno** [comment]
I added the macro (should we capitalize it?) and an use in the hello example.

I dont know how to add a test to check it works.. but I will figure something out tomorrow.

**Comment by zacharygrafton** [comment]
I would definitely capitalize the macro, but just because it is a typical convention, even for function call like macros, if it even needs to be a macro at all. I don't think it'd hurt if it was just an actual function call.

**Comment by davidmoreno** [comment]
I capitalized the macro names, and did some changes so the documentation is generated properly.

As the version.h.in is not recognized nor found by doxygen documentation has to be 100% on the .c file, including the macro documentation. 

The macros themselves need to be at the header file, where the ONION_VERSION_* are properly defined for currently running program, and not the compiled onion library, but with some tricks the documentation can be on the .c file.

**Comment by zacharygrafton** [comment]
Looks good to me. I'd go ahead and merge.

**Comment by davidmoreno** [comment]
I forgot to mention that the patch number is generated by git and it is the
number of commits since last tagging. This means it will not be serial per
patch released, but it will increase.

**Pull request merged**